### PR TITLE
feat: consume registrar's new mechanism of getting cram key using verification code

### LINKED ISSUE #330: Update onboarding_cli and activate_cli with new activation process of AtRegistrar
### Describe the bug

The registrar no longer supports the use of qr_code to activate an atsign. But onboarding_cli currently relies completely on qr_code /cram secret to activate an atsign. At the moment this makes onboarding_cli completely unusable for activation.

### Expected behavior

OnboardingCLI should be able to activate atsigns

### Additional context

AtRegistrar has migrated to an API call and OTP based cram_secret fetching instead of the having to use a qr_code. 

### Process Checklist

- [x] Modify OnboardingCLI activation process
- [x] Consume these changes in activate_cli
- [x] Update tests to match new behaviour
- [x] Update docs/readme
- [x] Publish new version of onboarding_cli

### PR DESCRIPTION
**Pull request by srieteja**
closes https://github.com/atsign-foundation/at_libraries/issues/330
**- What I did**
- rename RegisterUtil -> OnboardingUtil
- Introduced methods in OnboardingUtil that fetch cram key from registrar using verification code
- Consume verification code based atsign activation in AtOnboardingImpl
- Consume verification code based changes in activate_cli
- deprecate methods in AtOnboardingImpl that deal with qr_codes
- deprecate qr_code_path and introduce registrar_url in AtOnboardingPreference

**- How I did it**
- Introduced methods to make new API calls that fetch cram key from regstrar using verification code
- consume these methods in OnboardingCli and activate_cli
- deprecate all the methods/preferences that use qr_code

**- Description for the changelog**

feat: consume registrar's new mechanism of getting cram key using verification code

### CODE REVIEW

**Comment by cconstab** [comment]
Things do not work for me :-(

But once working this is going to be excellent !! TY


```
3 packages have newer versions incompatible with dependency constraints.
Try `dart pub outdated` for more information.
cconstab@tarial at_onboarding_cli % dart run bin/activate_cli.dart -a @blueheavybadger
[Information] Root server is root.atsign.org
[Information] Registrar url provided is my.atsign.com
[Information] Activating your atSign. This may take up to 2 minutes.
Successfully sent verification code to your registered e-mail
[Action Required] Enter your verification code:
66FT
[Error] Activation failed. Invalid data provided by user. Please try again
Cause: Please enter the 4-character verification code that was sent to your email address
cconstab@tarial at_onboarding_cli % 
```

Should I be using .wtf ? Or is this API fully working on prod? @athandle

**Comment by athandle** [comment]
@srieteja can you please check .
@cconstab APIs are working on prod as well

**Comment by srieteja** [comment]
@cconstab @athandle there was a typo which somehow creeped into the code and was causing the issue. I have now fixed it and have verified it to be properly working. Please try running the tool again to see how it works.
ps: I also upgraded the outdated dependencies

**Comment by gkc** [comment]
@srieteja Can you add some further test automation please so we don't have typos causing bugs undetected.

**Comment by srieteja** [comment]
> @srieteja Can you add some further test automation please so we don't have typos causing bugs undetected.

sure, @gkc

**Comment by cconstab** [comment]
Confirmed working (I also checked lower casing and that worked too) THANK YOU....

Let me know when pushed out to Pub.Dev as SSH No ports will depend on this

**Comment by srieteja** [comment]
Sure Colin, I'll push it out as soon as the tests are done.

**Comment by cconstab** [comment]
We agreed on the Arch call to push and work on the extra tests afterwards.. @murali-shris  has the details

**Comment by cconstab** [comment]
Could you look to fixing the issue mentioned in the ticket re trying to activate an already activated atSign. This will I think prevent confusion and problem tickets..

```
cconstab@tarial at_onboarding_cli % dart run bin/activate_cli.dart -a @blueheavybadger
[Information] Root server is root.atsign.org
[Information] Registrar url provided is my.atsign.com
[Information] Activating your atSign. This may take up to 2 minutes.
Successfully sent verification code to your registered e-mail
[Action Required] Enter your verification code: (verification code is not case-sensitive)
8xpq
[Information] CRAM Key fetched successfully
Connecting to secondary ...1/50
[Error] Activation failed. It looks like something went wrong on our side.
Please try again or contact support@atsign.com
Cause: Exception: Auth failed
cconstab@tarial at_onboarding_cli % 
```

**Comment by srieteja** [comment]
I cannot ascertain whether an atsign has already been activated or there's something else wrong. As both of these scenarios get the same response from registrar which says 'auth failed'. But we could introduce something in the registrar which would give a response saying 'atsign already activated' which I could use on the command-line tools. @athandle what do you think abou this ?

File: packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart [hunk 1]
```diff
@@ -81,17 +83,21 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
...
     //   return true;
     // }
-    //get cram_secret from either from AtOnboardingConfig or decode it from qr code whichever available
-    atOnboardingPreference.cramSecret ??=
-        getSecretFromQr(atOnboardingPreference.qrCodePath);
+
+    // get cram_secret from either from AtOnboardingPreference
+    // or fetch from the registrar using verification code sent to email
+    atOnboardingPreference.cramSecret ??= await OnboardingUtil()
+        .getCramUsingOtp(_atSign, atOnboardingPreference.registrarUrl);
```
Comment (murali-shris) [inline packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart:93]: temporarily write an example which tests this method call.

File: packages/at_onboarding_cli/lib/src/util/onboarding_util.dart [hunk 2]
```diff
@@ -127,33 +127,97 @@ class RegisterUtil {
...
+    }
+    throw at_client.InvalidRequestException(apiResponseMessage);
+  }
+
+  /// Returns the cram key for an atsign by fetching it from the registrar API
+  /// Accepts a registered [atsign], the verification code that was sent to
+  /// the registered email
+  /// Throws exception in the following cases:
+  /// 1) HTTP 400 BAD_REQUEST
+  Future<String> getAtsignCramKey(String atsign, String verificationCode,
```
Comment (murali-shris) [inline packages/at_onboarding_cli/lib/src/util/onboarding_util.dart]: rename this method to getCramKey

Comment (srieteja) [reply packages/at_onboarding_cli/lib/src/util/onboarding_util.dart]: change made as part of [5b38eea](https://github.com/atsign-foundation/at_libraries/pull/343/commits/5b38eeaadb5a861431de17602399df43c93ca0e4)

File: packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart (hunk 1 above)
Comment (srieteja) [reply packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart:93]: change made as part of [5b38eea](https://github.com/atsign-foundation/at_libraries/pull/343/commits/5b38eeaadb5a861431de17602399df43c93ca0e4)

**Comment by gkc** [comment]
I note that this will, technically, be a breaking change because you're renaming register_util, which is exported by the package.

Can you confirm please that other existing programs (e.g. at_talk) work unchanged with this new version of the package?

**Comment by srieteja** [comment]
> I note that this will, technically, be a breaking change because you're renaming register_util, which is exported by the package.
> 
> Can you confirm please that other existing programs (e.g. at_talk) work unchanged with this new version of the package?

I believe this will not break any At_Packages(at_talk, sshnoports, etc) as RegisterUtil class only contains methods related to RegisterCLI. Something like at_talk would only be using the AtOnboardingService which on an interface level has not changed. Although, if someone was using RegisterCli methods directly, it would break things for them. Should we consider tagging this release as a breaking change ?

**Comment by gkc** [comment]
No it's fine, as long as you've tested. Consider not exporting things that users of the package won't use directly.

File: packages/at_onboarding_cli/example/example.dart [hunk 3]
```diff
@@ -1,24 +1,47 @@
...
   AtLookUp? atLookup = onboardingService.atLookUp;
+  AtClient? client = onboardingService.atClient;
+  print(client?.getKeys());
   print(await atLookup?.scan(regex: 'publickey'));
   await onboardingService.close();
-  //free the object after it's used and no longer required
-  onboardingService = null;
+
+  //*****************************************************
+  // Example for onboarding an atsign with cram secret
```
Comment (murali-shris) [inline packages/at_onboarding_cli/example/example.dart]: move this block to separate example

File: packages/at_onboarding_cli/example/example.dart [hunk 4]
```diff
@@ -1,24 +1,47 @@
...
+      AtOnboardingServiceImpl('your atsign here', onboardingPreference);
+  await atOnboardingService.onboard(); // when authenticating
+  AtLookUp? atLookUp = onboardingService.atLookUp;
+  AtClient? atClient = onboardingService.atClient;
+  print(atClient?.getKeys());
+  print(await atLookUp?.scan(regex: 'publickey'));
+  await onboardingService.close();
+
+  //******************************************************
+  // Example to fetch cram key using verification code
```
Comment (murali-shris) [inline packages/at_onboarding_cli/example/example.dart]: move this block to separate example

File: packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart [hunk 5]
```diff
@@ -106,6 +117,8 @@ class AtOnboardingServiceImpl implements AtOnboardingService {
       if (_isAtsignOnboarded) {
         await _activateAtsign(atLookUpImpl);
       }
+    } on Exception catch (e) {
+      logger.info('Caught exception: $e');
```
Comment (murali-shris) [inline packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart]: change to logger.severe.
Change error message to 'Exception while activating atsign'
add a block for on Error as well

**Review by murali-shris — APPROVED** [review]

**Review by murali-shris — CHANGES_REQUESTED** [review]
requesting changes

File: packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart (hunk 5 above)
Comment (srieteja) [reply packages/at_onboarding_cli/lib/src/onboard/at_onboarding_service_impl.dart]: done as part of [bdf1a41](https://github.com/atsign-foundation/at_libraries/pull/343/commits/bdf1a41ca42fae375730f821e23c8e5f7269a127)

File: packages/at_onboarding_cli/example/example.dart (hunk 3 above)
Comment (srieteja) [reply packages/at_onboarding_cli/example/example.dart]: done as part of [b252c15](https://github.com/atsign-foundation/at_libraries/pull/343/commits/b252c15cab0d1d046ea394c7165c9ce935d5a877)

File: packages/at_onboarding_cli/example/example.dart (hunk 4 above)
Comment (srieteja) [reply packages/at_onboarding_cli/example/example.dart]: done as part of [b252c15](https://github.com/atsign-foundation/at_libraries/pull/343/commits/b252c15cab0d1d046ea394c7165c9ce935d5a877)

**Review by murali-shris — APPROVED** [review]

**Comment by murali-shris** [comment]
Approving this PR due to urgent client demo
Pending:
* functional tests for new changes
* review design of new changes

**Pull request merged**